zerver/models.py

```diff
@@ -2846,9 +2843,8 @@ def get_client_remote_cache(name: str) -> Client:
     return client
 
 
-@cache_with_key(get_stream_cache_key, timeout=3600 * 24 * 7)
 def get_realm_stream(stream_name: str, realm_id: int) -> Stream:
-    return Stream.objects.select_related().get(name__iexact=stream_name.strip(), realm_id=realm_id)
+    return Stream.objects.get(name__iexact=stream_name.strip(), realm_id=realm_id)
```

Comment:
Should we change this to `.select_related("realm")`. That version is safe from any of the weird recursive calls, and makes code that accesses `stream.realm` safe from extra queries... but I also don't know when we would want to do that, so maybe it's OK.

Reply:
I've audited that none of the calls via `access_stream_by_name` ever access `stream.realm` afterwards.

But I think the same won't be true when working with streams fetched by ID, and so we should pick a set of fields we want to promise that `Stream` objects have prefetched; and I think that should at least include `Recipient` (it's a tiny table that we haven't quite figured out how to inline into `Stream`). 

I'm open to skipping `Realm` -- I do think it's a bad pattern to access `stream.realm.foo`; `git grep stream[.]realm[.]` doesn't find a huge number of instances... but I think it might come with some risk, so I would argue for moving to `.select_related("realm", "recipient")` for stream objects as a low risk change we can just push through everywhere.

Reply:
I am pretty adamant that we should stick to Stream-only objects.  I accept the fact that `Recipient` is tiny, but I know that I have done a lot of work in the past to make sure that we don't erroneously try to use it for things are completely implied by Stream.  Stream has recipient_id and the Recipient.type is implied to be Recipient.STREAM.

We have enough test coverage of query counts on all our major things to be sure that we're not hurting things like page-load or message-sending insofar as we make this mistake at all.

Reply:
Thinking about this a more, the detail I was missing is that `stream.recipient_id` lets us get the only field on `Recipient` that we need without actually following the join.

Reply:
OK, I think I've convinced myself that any code using a Stream object accessing `stream.realm.something` or `stream.recipient.something` is going to be code that doesn't make sense to be written that way anyway.